tools/base/parallel.py

```diff
@@ -0,0 +1,74 @@
+
+import asyncio
+import argparse
+import math
+import os
+import sys
+from functools import cached_property
+
+import abstracts
+
+from aio.core.event import AExecutive, IExecutive
+from aio.run import runner
+
+
+@abstracts.implementer(IExecutive)
```

Comment:
@lizan @keith is this reasonable to be doing or is this an anti-pattern? Basically, we're building a parallel execution engine within a Bazel action? Intuitively I would think we would want to rely on Bazel itself to provide the scheduling.

Reply:
please note this code is no longer part of this PR - the code was moved upstream here - https://github.com/envoyproxy/pytooling/blob/main/envoy.base.utils/envoy/base/utils/parallel_runner.py

i should also note that both the code.check and dependency.check tools effectively orchestrate parallel procs - its this that has made them much faster

implementing this in bazel should be possible - but as commented elsewhere i dont think you can get the number of available procs _for_ tasks only _in_ tasks

Reply:
I think there's some practical balance between bazel parallelizing vs it exec'ing tools that then parallelize. If the common use case was to run N tools in bazel at the same time, and they also fanned out, that could cause issues, but I've mostly found in the common case you'll just exec 1 or 2 tools that can parallelize and I think that's alright.

Reply:
> ... run N tools in bazel at the same time, and they also fanned out, that could cause issues

yep, agreed that is a concern - the forking procs are essentially not using bazel's proc pool

that said, i think its mostly a theoretical concern, OSs are pretty good at managing simultaneous multicore procs these days

in this case the amount of time spent in my current target (building `//docs:api_rst`) falls from ~7mins on 4 cores -> ~1.20s running serially

that suggests there is a lot less work being done overall, by a pretty big margin

pushing the new code onto 4 cores takes the build time down <30s

Reply:
Well, it's not just CPU scheduling, it's also the ability to do incremental doc builds, right? If Bazel is aware of inputs to individual RST builds, we can avoid rebuilding, so the incremental iteration case becomes faster.

Reply:
related:

- https://github.com/Ubehebe/bazel-worker-examples/blob/master/echo.py#L24
- https://github.com/Ubehebe/bazel-worker-examples/blob/master/forked/BUILD#L8

Reply:
- https://github.com/grpc/grpc/blob/master/tools/distrib/python/grpcio_tools/grpc_tools/main.cc

Reply:
i have a reasonable implementation of a stdin <> stdout line receiver that i was working on previously (https://github.com/envoyproxy/pytooling/pull/273)

im going to try:

- add protobuf Worker parsing
- calling grpcio tools to run a "protoc" like command in python
- setup above as persistent worker

if that works i think we will get the best of all worlds - aspects, incremental builds, bazel defined parallelism, efficient parsing of compile data

Reply:
i have a vaguely working protoc-plugin-compiler-as-a-service

i wasnt able to get it working using protoc itself via `grpc_tools` - altho they wrap the c++ lib they call `cli.Run(argc, etc)` so afaict there is not a way to use it without hitting the same problem of reparsing the descriptor file on every invokation

i think it does a lot of things we dont want anyway

what i was able to do was read the proto file source from a descriptor set and feed that directly to the api plugin

i still need to hook up the bazel service and above needs a bit of work yet but i think its pretty doable

Reply:
the other downside to the approach here (so far) is that it is using the compiled protoc 

i havent got the new plan quite working yet, but my initial r&d suggests that cutting this out - using the `protoc` in `rules_proto` to derive the initial descriptor, and then using python to generate the plugin/rst output speeds things up considerably

i need to confirm everything yet, but its looking like we are going from ~7min -> ~20s for api rst build !!!!

Reply:
after getting the aspect + worker working it takes what was originally a ~7min job (rebuilding `api_rst` after python change, so all api files) down to ~15s

wrt re/compiling `protoc` this is still an issue i think, but a separate one

im going to close this PR in favour of #21671 given that it is both faster and preferable for the reasons discussed above